### PR TITLE
[InstCombine] Generalize `icmp (shl nuw C2, Y), C -> icmp Y, C3`

### PR DESCRIPTION
**Pull request by dtcxzyw**
The motivation of this patch is to fold more generalized patterns like `icmp ult (shl nuw 16, X), 64 -> icmp ult X, 2`.

Alive2: https://alive2.llvm.org/ce/z/gyqjQH

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 1]
```diff
@@ -2227,18 +2227,24 @@ Instruction *InstCombinerImpl::foldICmpMulConstant(ICmpInst &Cmp,
...
 
   Type *ShiftType = Shl->getType();
   unsigned TypeBits = C.getBitWidth();
-  bool CIsPowerOf2 = C.isPowerOf2();
   ICmpInst::Predicate Pred = Cmp.getPredicate();
   if (Cmp.isUnsigned()) {
+    APInt Div, Rem;
+    APInt::udivrem(C, *C2, Div, Rem);
+    if (!Rem.isZero())
+      return nullptr;
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:2245]: Can this be generalized to the non-rem-0 case? Taking your example, `icmp ult (shl nuw C2, X), 64 -> icmp ult X, 2` holds for any C2 in [16,31].

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:2245]: All non-rem-0 cases:
```
Input files: 37900
Progress: 3183icmp ugt (shl nuw 12, X), 72057594037927935
icmp ugt (shl nuw 12, X), 72057594037927935
Progress: 3189icmp ugt (shl nuw 12, X), 72057594037927935
Progress: 3279icmp ugt (shl nuw 12, X), 72057594037927935
Progress: 3308icmp ugt (shl nuw 12, X), 72057594037927935
icmp ugt (shl nuw 12, X), 72057594037927935
Progress: 6875icmp ugt (shl nuw 4, X), 31
Progress: 16038icmp ugt (shl nuw 2, X), 134217727
icmp ugt (shl nuw 2, X), 134217727
icmp ugt (shl nuw 2, X), 134217727
icmp ugt (shl nuw 2, X), 134217727
icmp ugt (shl nuw 2, X), 134217727
icmp ugt (shl nuw 2, X), 134217727
icmp ugt (shl nuw 2, X), 134217727
icmp ugt (shl nuw 2, X), 134217727
Progress: 18015icmp ugt (shl nuw 16, X), 63
Progress: 24392icmp ugt (shl nuw 2, X), 3
Progress: 37893
8
c3c/optimized/target.c.ll
folly/optimized/DynamicParser.cpp.ll
folly/optimized/LogConfigParser.cpp.ll
folly/optimized/dynamic.cpp.ll
folly/optimized/json.cpp.ll
linux/optimized/xhci-pci.ll
regex-rs/optimized/11vfjke4utuj478u.ll
wasmtime-rs/optimized/2ly4gzztxx8hlwxv.ll
```

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:2245]: > Can this be generalized to the non-rem-0 case? Taking your example, `icmp ult (shl nuw C2, X), 64 -> icmp ult X, 2` holds for any C2 in [16,31].

Do you have any ideas to handle these non-rem-0 cases? If not, I will convert `icmp ugt (shl nuw C2, X), C3 -> icmp uge (shl nuw C2, X), C3 + 1` to cover more real-world cases.

**Comment by nikic** [comment]
> Alive2: https://alive2.llvm.org/ce/z/0xcdE6

These proofs don't match the implementation. Note that the pow2 check is not a pre-condition of the transform, it is only used to adjust the predicate in some cases. There is no adjustment for ule predicates, so in that case I'd expect this to verify for the implementation to be correct: https://alive2.llvm.org/ce/z/wk4LYe

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp (hunk 1 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:2245]: Isn't it just find the first X value where `C2 nuw<< X u< C1` is false? Which is just `log2(roundup_next_p2(C1 / C2))` no?

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp]: Done. Thank you!

**Comment by dtcxzyw** [comment]
> > Alive2: https://alive2.llvm.org/ce/z/0xcdE6
> 
> These proofs don't match the implementation. Note that the pow2 check is not a pre-condition of the transform, it is only used to adjust the predicate in some cases. There is no adjustment for ule predicates, so in that case I'd expect this to verify for the implementation to be correct: https://alive2.llvm.org/ce/z/wk4LYe

Updated.

**Comment by dtcxzyw** [comment]
Ping.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 2]
```diff
@@ -2251,9 +2257,9 @@ static Instruction *foldICmpShlOne(ICmpInst &Cmp, Instruction *Shl,
         Pred = ICmpInst::ICMP_UGT;
     }
 
-    unsigned CLog2 = C.logBase2();
+    unsigned CLog2 = Div.logBase2();
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:2260]: Can you add a test where `C` and `C2` are not a power of two and we need to roundup the log? I.e
`(48 << X) u>= 144`?

I don't see the ceil logic for that case.

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:2260]: https://alive2.llvm.org/ce/z/XejB_d

**Review by goldsteinn — COMMENTED** [review]
LGTM give nikic a day before pushing.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 3]
```diff
@@ -2227,18 +2227,24 @@ Instruction *InstCombinerImpl::foldICmpMulConstant(ICmpInst &Cmp,
...
+  if (!match(Shl, m_NUWShl(m_APInt(C2), m_Value(Y))))
     return nullptr;
 
   Type *ShiftType = Shl->getType();
   unsigned TypeBits = C.getBitWidth();
-  bool CIsPowerOf2 = C.isPowerOf2();
   ICmpInst::Predicate Pred = Cmp.getPredicate();
   if (Cmp.isUnsigned()) {
+    assert(!C2->isZero() && C2->ule(C) &&
+           "Should be simplified by InstSimplify");
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:2243]: I'd rather not make this an assert. Note that you can only rely on the root instruction being simplified. Other instructions may not be simplified due to weird worklist order.

File: llvm/test/Transforms/InstCombine/icmp-shl-nuw.ll [hunk 4]
```diff
@@ -90,3 +90,109 @@ define <2 x i1> @icmp_ugt_16x2(<2 x i32>) {
...
+; CHECK-LABEL: @fold_icmp_shl_nuw_c3(
+; CHECK-NEXT:    [[CMP:%.*]] = icmp ugt i32 [[X:%.*]], 1
+; CHECK-NEXT:    ret i1 [[CMP]]
+;
+  %shl = shl nuw i32 48, %x
+  %cmp = icmp uge i32 %shl, 144
+  ret i1 %cmp
+}
+
+; Negative tests
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/icmp-shl-nuw.ll:147]: The first test here is no longer a negative test?

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**